File: golang/behavioral/observer/observer.go
/*!
* \brief   Observer. Define a one-to-many dependency
*        between objects so that when one object changes
*        state,all its dependents are notified and updated
*        automatically.
 */

package main

import "fmt"

type Subject struct {
	observer_list []Observer
	status        int
}

func CreateSubject() *Subject {
	return &Subject{
		observer_list: make([]Observer, 0),
	}
}

func (s *Subject) Add(o Observer) {
	s.observer_list = append(s.observer_list, o)
}

func (s *Subject) Remove(i int) {
	s.observer_list = append(s.observer_list[:i], s.observer_list[i+1:]...)
}

func (s *Subject) notify() {
	for _, o := range s.observer_list {
		o.Update()
	}
}

func (s *Subject) SetState(status int) {
	s.status = status
	s.notify()
}

func (s *Subject) GetStatus() int {
	return s.status
}

/////////////

type Observer interface {
	Update()
	GetStatus() int
}

type ConcreteObserver struct {
	subject *Subject
	status  int
}

func CreateObserver(subject *Subject) *ConcreteObserver {
	return &ConcreteObserver{
		subject: subject,
	}
}

func (co *ConcreteObserver) Update() {
	co.status = co.subject.GetStatus()
}

func (co *ConcreteObserver) GetStatus() int {
	return co.status
}

func main() {
	subject := CreateSubject()
	observer1 := CreateObserver(subject)
	observer2 := CreateObserver(subject)

	subject.Add(observer1)
	subject.Add(observer2)

	fmt.Printf("subject: %d, observer1: %d, observer2: %d.\n",
		subject.GetStatus(), observer1.GetStatus(), observer2.GetStatus())

	subject.SetState(5)

	fmt.Printf("subject: %d, observer1: %d, observer2: %d.\n",
		subject.GetStatus(), observer1.GetStatus(), observer2.GetStatus())

	subject.SetState(10)

	fmt.Printf("subject: %d, observer1: %d, observer2: %d.\n",
		subject.GetStatus(), observer1.GetStatus(), observer2.GetStatus())
}

File: design-patterns/cpp/behavioral/observer.cpp
/*!
* \brief   Observer. Define a one-to-many dependency 
*        between objects so that when one object changes
*        state,all its dependents are notified and updated
*        automatically.
*
*          Subject      <->       Observer
*             |                      |                        
*     ConcreteSubject          ConcreteObserver  
*/

#include <iostream>
#include <list>

class Observer {
public:
  virtual void Update() = 0;
  virtual int GetState() = 0;
};

class Subject {
public:
  virtual void Add(Observer *) = 0;
  virtual void Remove(Observer *) = 0;
  virtual void Notify() = 0;
  virtual void SetState(int state) = 0;
  virtual int GetState() = 0;
};

class ConcreteObserver : public Observer {
public:
  ConcreteObserver(Subject *subject) : subject_(subject), status_(0){}

  void Update() {
    status_ = subject_->GetState();
  }
  int GetState() {
    return status_;
  }

private:
  Subject *subject_;
  int status_;
};

class ConcreteSubject : public Subject {
public:
  ConcreteSubject() : status_(0) {}
  void Add(Observer *observer) {
    observer_list_.push_back(observer);
  }
  void Remove(Observer *observer) {
    observer_list_.remove(observer);
  }
  void Notify() {
    std::list<Observer *>::iterator it = observer_list_.begin();
    while (it != observer_list_.end()) {
      (*it)->Update();
      ++it;
    }
  }

  void SetState(int state) {
    status_ = state;
  }
  int GetState() {
    return status_;
  }
private:
  std::list<Observer *> observer_list_;
  int status_;
};

int main() {
  // Create Subject
  Subject *subject = new ConcreteSubject();

  // Create Observer
  Observer *observer1 = new ConcreteObserver(subject);
  Observer *observer2 = new ConcreteObserver(subject);
 
  printf("subject: %d, observer1: %d, observer2: %d.\n", 
    subject->GetState(), observer1->GetState(), observer2->GetState());
  // Register the observer
  subject->Add(observer1);
  subject->Add(observer2);

  // Change the state
  subject->SetState(2);
  subject->Notify();

  // The status of the observer changes as the subject changes.
  printf("subject: %d, observer1: %d, observer2: %d.\n",
    subject->GetState(), observer1->GetState(), observer2->GetState());

  // Unregister the observer
  subject->Remove(observer2);

  subject->SetState(3);
  subject->Notify();

  printf("subject: %d, observer1: %d, observer2: %d.\n",
    subject->GetState(), observer1->GetState(), observer2->GetState());

  delete observer1;
  delete observer2;
  delete subject;
}